File: Interpretadas/Python/radix_sort.py
import sys
import time
import random
import locale

# Classe para armazenar as métricas
class Metricas:
    def __init__(self):
        self.comparacoes = 0  # Comparacoes não são feitas diretamente no Radix Sort
        self.trocas = 0  # Trocas não são feitas diretamente no Radix Sort
        self.tempoExecucao = 0.0  # Armazena o tempo de execução em segundos
        self.memoriaUsada = 0  # Armazena a quantidade de memória usada em bytes

# Implementação do Counting Sort usado pelo Radix Sort
def countingSortForRadix(arr, exp, metricas):
    n = len(arr)
    output = [0] * n
    count = [0] * 10
    metricas.memoriaUsada += (len(count) + len(output)) * 4  # Considerando 4 bytes por int

    for i in range(n):
        index = arr[i] // exp
        count[index % 10] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]

    for i in range(n - 1, -1, -1):
        index = arr[i] // exp
        output[count[index % 10] - 1] = arr[i]
        count[index % 10] -= 1

    for i in range(n):
        arr[i] = output[i]

# Implementação do Radix Sort
def radixSort(arr, metricas):
    max_val = max(arr)
    exp = 1
    while max_val // exp > 0:
        countingSortForRadix(arr, exp, metricas)
        exp *= 10


def carregarVetor(tamanho, caso):
    with open('vetores/vetores_input_2.txt', 'r') as f:
        lines = f.readlines()
        key = f"{tamanho} {caso}:"
        for i in range(len(lines)):
            if lines[i].strip() == key:
                vetor = list(map(int, lines[i + 1].strip().split()))
                return vetor
    return []

# Função principal
def main(tamanho, caso):
    locale.setlocale(locale.LC_ALL, "Portuguese")

    # N = 100000  # Tamanho do vetor
    # opcao = 0  # 0 para aleatório, 1 para crescente, 2 para decrescente
    # repeticao = 1  # 0 para sem repetição, 1 para com repetição

    # # Testando Radix Sort
    # metricas = Metricas()
    # arr_radix = (N, opcao, repeticao)
    # inicio = time.time()
    # radixSort(arr_radix, metricas)
    # fim = time.time()
    # metricas.tempoExecucao = fim - inicio
    


    

    metricas = Metricas()
    arr = carregarVetor(tamanho, caso)
    
    inicio = time.time()
    radixSort(arr, metricas)
    fim = time.time()
    metricas.tempoExecucao = fim - inicio

    # Retornar métricas como dicionário
    return {
        "Comparacoes": metricas.comparacoes,
        "Trocas": metricas.trocas,
        "Tempo de execucao": metricas.tempoExecucao,
        "Memoria usada": metricas.memoriaUsada
    }

if __name__ == "__main__":
    main()

File: Interpretadas/Javascript/radix_sort.js
const fs = require('fs');
const { performance } = require('perf_hooks');

class Metricas {
    constructor() {
        this.comparacoes = 0; // Comparacoes não são feitas diretamente no Radix Sort
        this.trocas = 0;      // Trocas não são feitas diretamente no Radix Sort
        this.tempoExecucao = 0.0; // Em segundos
        this.memoriaUsada = 0;    // Em bytes
    }
}

function countingSortForRadix(arr, exp, metricas) {
    const n = arr.length;
    const output = new Array(n).fill(0);
    const count = new Array(10).fill(0);

    metricas.memoriaUsada += (count.length + output.length) * 4;

    for (let i = 0; i < n; i++) {
        const index = Math.floor(arr[i] / exp) % 10;
        count[index]++;
    }

    for (let i = 1; i < 10; i++) {
        count[i] += count[i - 1];
    }

    for (let i = n - 1; i >= 0; i--) {
        const index = Math.floor(arr[i] / exp) % 10;
        output[count[index] - 1] = arr[i];
        count[index]--;
    }

    for (let i = 0; i < n; i++) {
        arr[i] = output[i];
    }
}

function radixSort(arr, metricas) {
    const max_val = Math.max(...arr);
    for (let exp = 1; Math.floor(max_val / exp) > 0; exp *= 10) {
        countingSortForRadix(arr, exp, metricas);
        metricas.Comparacoes++;
    }
}

function carregarVetor(tamanho, caso) {
    const data = fs.readFileSync('D:/Documentos/cefet/AEDS/Trabalho_3/vetores/vetores_input_2.txt', 'utf8');
    const lines = data.split('\n');
    const key = `${tamanho} ${caso}:`;
    for (let i = 0; i < lines.length; i++) {
        if (lines[i].trim() === key) {
            return lines[i + 1].trim().split(' ').map(Number);
        }
    }
    return [];
}

function main(tamanho, caso) {
    const metricas = new Metricas();
    const arr = carregarVetor(tamanho, caso);
    
    const inicio = performance.now();
    radixSort(arr, metricas);
    const fim = performance.now();
    metricas.tempoExecucao = (fim - inicio) / 1000;

    console.log({
        "Comparacoes": metricas.comparacoes,
        "Trocas": metricas.trocas,
        "Tempo de execucao": metricas.tempoExecucao,
        "Memoria usada": metricas.memoriaUsada
    });
}

// Exemplo de chamada
// main(1000, "aleatorios");
